Waypoint types

lhelwerd opened the pull request:
Implement three waypoint types: setting the vehicle's **home location** at the start of the mission, moving it via a waypoint that it is allowed to **pass**, and going to a waypoint to **wait** for certain other vehicles.

These types are shown in the control panel waypoints view and used in the RF sensor mission. The waypoints view was updated to perform more constraint validity, as well as other cleanup to the cell interaction such as keyboard navigation. The planning view generates valid wait type waypoints, but could later maybe provide other information.

The RF sensor mission uses the waypoint types to offload the waypoint generation and deduces whether we should add a wait waypoint to the mission.

Other waypoint types may be added later, and they might be used in other missions to generate patterns such as the fan beam, straight line and calibration cycles. This is not part of this PR, which intends to implement the basic requirements for different waypoint types.

lhelwerd committed:
Control panel Waypoints view: Add waypoint type

- Add a column for selecting the type of a waypoint. Each cell contains
  a selector which can be set to "home", "pass" or "wait". These
  correspond to certain stub waypoint types.
- Refactor the handling of other columns to be more sensible.
- Keep the differing types into account when compressing and
  uncompressing.
- Add backward compatibility for older JSON files.
- The selector disables the wait ID and wait count columns when the type
  is not "wait". It is visually greyed out, no longer selectable through
  tabbing and its value is considered to be default when retrieving it
  for exports and so on.
- Change tab ordering such that we tab over the same row but skip over
  the columns with defaults that are `0`, i.e., altitude and wait ID.
  Also skip over disabled cells instead of getting stuck there. This
  makes it easier to insert values for the type and wait count. Open an
  editor in the selected cell if possible, to improve edit efficiency.
- Add some documentation and create the stub waypoint module.
- Fix state after reload. The problem is that when the waypoint data is
  imported or reloaded, the rows are removed and then readded, but
  a default column is removed at that point. Refactor this so that we do
  not need to remove those item widgets anymore.

lhelwerd committed:
Waypoints: Add "type" to packet, update interface

- Send the type in the "waypoint_add" packet, and update the tests and
  specifications so that we have the type.
- Add a create class method to `Waypoint`, and track its location. This
  should make it easier to set them up and implement the waypoint types.

lhelwerd committed:
Mission RF Sensor: Implement wait type waypoint

- Replace the range generation in `Mission_RF_Sensor` with the use of
  waypoint types, which provide us with the waypoints and required
  sensors. Currently, `Waypoint_Wait` is the only supported waypoint
  type, and any other type would cause several different exceptions due
  to loading and retrieving required sensors, but these are implemented
  in a followup commit.
- Mention the enum34 backward compatibility module as a dependency.
- Ensure that the `Mission_RF_Sensor` has a RF sensor, since it might be
  disabled via settings which makes this mission useless.
- Test the new classes and functionality.

lhelwerd committed:
Waypoint types: Add pass waypoint

This (re)introduces the possibility to add passable waypoints to an
automated mission. This can be useful to require the vehicle to move
around to another waypoint in a certain manner. This we can perform
measurements at waiting locations and then quickly move to faraway
location via a passable waypoint, which is helpful for creating safe
missions during the planning phase, for example.

The `Mission_RF_Sensor` now correctly distinguishes between wait
waypoints that have required sensors, wait waypoints that want to wait
for all other sensors and waypoints that do not want to wait at all.

lhelwerd committed:
Control panel Waypoints: Enforce minimal values

The wait ID should be at least 0, since any lower value has no meaning.
The wait count should be at least 1, because a value of zero causes the
range of waypoints to be empty. Provide an error when this is the case
during exporting/sending/compressing, and display invalid columns
visually using a background color. This makes it easier to spot problems
such as values that cannot be casted, or that are too low.

lhelwerd committed:
Waypoint types: implement home waypoint

The home waypoint type immediately changes the home location of the
vehicle. It may thus only occur once at the beginning of the waypoints
in the control panel.

Add a method to the `Waypoint` interface to be called after adding the
command, so that the waypoint can update other vehicle state. This is
used to update the home location, and tests show it works as expected.

Other bug fixes in the control panel:
- Remove a method from the waypoints view that was moved to the table
  widget, but its calls were not yet updated.
- Make the waypoint type widget track the item it belongs to, because
  the row (and theoretically the column) could change after the object
  is created, e.g., when we insert or delete a row before this one.
  Create a method to retrieve the item at a certain position, or create
  it if it did not exist already, so that we can ensure that the item
  exists when we create the cell widget.

lhelwerd committed:
Planning: Correct waypoint type in export, update defaults

- Fill the assignment with enough data that the waypoints view needs.
- Make discrete planning default and update default home location based
  on grid network size.
- Update tests

lhelwerd committed:
Control panel Waypoints: Add more comments

timvandermeij commented:
Since this changes mission code, we should probably test this on the hardware too before merging.

lhelwerd committed:
RF sensor mission: Send acknowledgement after done packet

We need to ensure that the done packet arrives, otherwise the vehicle
does not enter the correct state. Thus handle the acknowledgement
correctly.
